### PR TITLE
Cosmos Spark - switching to reflection instead of bridge-approach to avoid SecurityException when installing connector by directly copying into /databricks/jars folder

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description
When deploying the Connector in Databricks by copying it directly into /databricks/jars folder instead of using the Databricks API or UI to install the connector a SecurityException (see below) would be thrown for write operations.

```
java.lang.SecurityException: class "org.apache.spark.SparkInternalsBridge$"'s signer information does not match signer information of other classes in the same package
```

This PR switches from using a bridge-approach to using reflection. It also allows opting-out of providing custom metrics in case there are any issues with using the reflection code. And for any reflection related failures custom metrics will simply be suppressed.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      Option.apply(TaskContext.get()) match {
+        case Some(taskCtx) => getInternalCustomTaskMetricsAsSQLMetricInternal(knownCosmosMetricNames, taskCtx)
+        case None => Map.empty[String, SQLMetric]
+      }
+    }
+  }
+
+  private def getAccumulators(taskCtx: TaskContext): Seq[AccumulatorV2[_, _]] = {
+    val taskMetrics:Object = taskCtx.taskMetrics()
+    taskMetrics.getClass.getMethod("accumulators").invoke(taskMetrics).asInstanceOf[Seq[AccumulatorV2[_, _]]]
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:54]: do we need to make the accumulators method accessible?` (accumulators.setAccessible(true))`

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          val sqlMetric = accumulable.asInstanceOf[SQLMetric]
+          sqlMetric.name.get -> sqlMetric
+        })
+        .toMap[String, SQLMetric]
+    } catch {
+      case e: Exception => {
+        logInfo(s"Could not populate custom metrics - Error ${e.getMessage}", e)
+
+        // reflection failed - disabling it for the future
+        reflectionAccessAllowed.set(false)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:77]: the exception being captured here is generic errors, should we only disallow for `InvocationTargetException` or similar?

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:54]: But I guess we should - I also want to cache the reflection logic

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:54]: Not ready for review yet - forgot to set the draft mode

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      }
+
+      val outputMetrics = method.invoke(taskMetrics)
+      Option(outputMetrics)
+    } catch {
+      case e: Exception =>
+        logInfo(s"Could not invoke getOutputMetrics via reflection - Error ${e.getMessage}", e)
+
+        // reflection failed - disabling it for the future
+        reflectionAccessAllowed.set(false)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:65]: should we only disable the reflection access for `InvocationTargetException`

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala]: Keep ctaching generic exception intentionally - but scoped the try-block to the relevant reflection code.

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:65]: See above

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}
+
+//scalastyle:off multiple.string.literals
+object SparkInternalsBridge extends BasicLoggingTrait {
+  private val SPARK_REFLECTION_ACCESS_ALLOWED_PROPERTY = "COSMOS.SPARK_REFLECTION_ACCESS_ALLOWED"
+  private val SPARK_REFLECTION_ACCESS_ALLOWED_VARIABLE = "COSMOS_SPARK_REFLECTION_ACCESS_ALLOWED"
+
+  private val DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED = true
+
+  val NUM_ROWS_PER_UPDATE = 100
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:21]: where this val is used?

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala [hunk 5]
```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.implementation.guava25.base.MoreObjects.firstNonNull
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala:5]: changelog entry~

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Pull request merged**